### DISCUSSION TITLE
OSS projects restricted from code scanning REST APIs

### DISCUSSION THREAD

**Discussion by 0x73746F66** [community/community#128794] [Apps, API and Webhooks]
### Select Topic Area

Bug

### Body

I can see everything is working in [code-scanning CodeQL configurations](https://github.com/trivialsec/triage-by-trivial-security/security/code-scanning/tools/CodeQL/status/configurations/automatic) and [rules](https://github.com/trivialsec/triage-by-trivial-security/security/code-scanning/tools/CodeQL/status/rules.csv?ids%5B%5D=233521077) are easy to view too.

I can see [github-advanced-security](https://github.com/trivialsec/triage-by-trivial-security/actions/runs/9533356993) doing it's thing, it is an OSS repo so it's a really nice feature.

But how to obtain the SARIF?
The gha has no artifacts, the logs don't indicate SARIF creation at all.
I do see [reporting](https://github.com/trivialsec/triage-by-trivial-security/pull/12/checks?check_run_id=26290392653) exists, not sure what that means if it is SARIF or not..

I tied to locate the API and believe I [have located it](https://docs.github.com/en/rest/code-scanning/code-scanning?apiVersion=2022-11-28#list-code-scanning-analyses-for-a-repository) 

```
*   Trying 20.248.137.49:443...
* Connected to api.github.com (20.248.137.49) port 443
* ALPN: curl offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (IN), TLS handshake, Server hello (2):
* (304) (IN), TLS handshake, Unknown (8):
* (304) (IN), TLS handshake, Certificate (11):
* (304) (IN), TLS handshake, CERT verify (15):
* (304) (IN), TLS handshake, Finished (20):
* (304) (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / AEAD-CHACHA20-POLY1305-SHA256
* ALPN: server accepted h2
* Server certificate:
*  subject: CN=*.github.com
*  start date: Mar  7 00:00:00 2024 GMT
*  expire date: Mar  7 23:59:59 2025 GMT
*  subjectAltName: host "api.github.com" matched cert's "*.github.com"
*  issuer: C=GB; ST=Greater Manchester; L=Salford; O=Sectigo Limited; CN=Sectigo ECC Domain Validation Secure Server CA
*  SSL certificate verify ok.
* using HTTP/2
* [HTTP/2] [1] OPENED stream for https://api.github.com/repos/trivialsec/triage-by-trivial-security/code-scanning/analyses?per_page=100&page=1
* [HTTP/2] [1] [:method: GET]
* [HTTP/2] [1] [:scheme: https]
* [HTTP/2] [1] [:authority: api.github.com]
* [HTTP/2] [1] [:path: /repos/trivialsec/triage-by-trivial-security/code-scanning/analyses?per_page=100&page=1]
* [HTTP/2] [1] [accept: application/vnd.github+json]
* [HTTP/2] [1] [authorization: token ghu_<redacted>]
* [HTTP/2] [1] [x-github-api-version: 2022-11-28]
* [HTTP/2] [1] [user-agent: Triage-by-Trivial-Security]
> GET /repos/trivialsec/triage-by-trivial-security/code-scanning/analyses?per_page=100&page=1 HTTP/2
> Host: api.github.com
> Accept: application/vnd.github+json
> Authorization: token ghu_<redacted>
> X-GitHub-Api-Version: 2022-11-28
> User-Agent: Triage-by-Trivial-Security
>
< HTTP/2 403
< server: GitHub.com
< date: Mon, 17 Jun 2024 00:19:10 GMT
< content-type: application/json; charset=utf-8
< content-length: 191
< x-oauth-scopes:
< x-accepted-oauth-scopes: admin:repo_hook, delete_repo, read:repo_hook, repo, repo:invite, repo:status, repo_deployment, security_events, write:repo_hook
< x-oauth-client-id: Iv23liW5R5lkjMRgFrWI
< x-github-media-type: github.v3; format=json
< x-accepted-github-permissions: security_events=read
< x-github-api-version-selected: 2022-11-28
< x-ratelimit-limit: 5000
< x-ratelimit-remaining: 4999
< x-ratelimit-reset: 1718587150
< x-ratelimit-used: 1
< x-ratelimit-resource: core
< access-control-expose-headers: ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset
< access-control-allow-origin: *
< strict-transport-security: max-age=31536000; includeSubdomains; preload
< x-frame-options: deny
< x-content-type-options: nosniff
< x-xss-protection: 0
< referrer-policy: origin-when-cross-origin, strict-origin-when-cross-origin
< content-security-policy: default-src 'none'
< vary: Accept-Encoding, Accept, X-Requested-With
< x-github-request-id: E7BA:1DFC62:CF049A:DA67C0:666F80FE
<
* Connection #0 to host api.github.com left intact
```
```json
{"message":"Resource not accessible by integration","documentation_url":"https://docs.github.com/rest/code-scanning/code-scanning#list-code-scanning-analyses-for-a-repository","status":"403"}
```

So that `403` Status code is documented as `Response if GitHub Advanced Security is not enabled for this repository` Which makes no sense as this is clearly enabled for this repository..

I expected to get this response and retrieve the `id` to follow up with a call to the [SARIF API](https://docs.github.com/en/rest/code-scanning/code-scanning?apiVersion=2022-11-28#get-a-code-scanning-analysis-for-a-repository) using `application/sarif+json`.

Because 403 might be due to the GitHub App I checked the permissions:
![app permissions](https://github.com/community/community/assets/93355168/c282fc3b-3d60-4a9f-b0e7-0568f8df5287)

And the accepted perms are accepted too, including `code` for code scanning:
![accepted perms](https://github.com/community/community/assets/93355168/9412f5ba-058a-44de-b486-64950bc4d4f0)

The only thing I can assume here is the API is still coded based on the days when code scanning was an enterprise plan only, and therefore the API is buggy for OSS repositories.

As an OSS project, I won't be able to _tell someone_ they have a bug - I am very confident that is the root cause.

If I missed something I am open minded, but it is so far very clearly NOT working as designed due to legacy plan restrictions.

**Comment by bogdanap** [comment]
Hi @0x73746F66.

Thanks for reaching out! 
While I agree the docs can be a bit confusing, HTTP 403 is also returned when the authentication token does not have the required permission to access the resources. 
This seems to be the case based on the response you received  (`{"message":"Resource not accessible by integration"}`).

The REST API is accessible for public repos and there should not be any plan restrictions. 
You just need to make sure the token you're using has the required `Code Scanning alerts` repository permission and you are correctly [authenticating as a GitHub App installation](https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/authenticating-as-a-github-app-installation) before issuing the request.

**Reply by 0x73746F66** [reply]
The post shows the permissions, which is clearly why I made the assertion you attempted to correct but obviously didn't notice the advice you gave was already considered in my report.

**Reply by aegilops** [reply]
Hi Stof. I created an GitHub App just now to test what the permissions are shown as, since I was pretty sure that "code" is _not_ what the "Code scanning alerts" permission correlates to.

What I found doesn't solve the problem, but it at least clears up which permission we're talking about.

"code" on that screen you showed demonstrates that the app has the "Contents" access permission. Access to the "Code scanning alerts" permission is reflected as "security events" on that screen.

You _do_ also have "security events" shown, which _should_ be sufficient, but as you say there's still a problem.

That leaves the question of how you are generating the access token for the GitHub App.

Can you try falling back to a fine-grained PAT and checking that it works as expected please? If it does, that rules out something with the permissions of the Code Scanning REST API on public repos. It it doesn't work, then please let us know here.

If the PAT works, then please debug your generation and use of [authenticating as a GitHub App installation](https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/authenticating-as-a-github-app-installation), as Bogdana suggested.

**Reply by 0x73746F66** [reply]
As shown, it is a GitHub App
i.e. a PAT is not something that makes sense at all to test.
I appreciate the testing, however this last request does not make any sense.

**Reply by aegilops** [reply]
I won't be engaging further with you

**Reply by 0x73746F66** [reply]
Thanks for attempting.
Sorry I couldn't test a PAT, the GitHub App does not get any PAT, it can't.

Who takes bugs, clear bugs like this, and gets them reported?
